File: 0940-fruit-into-baskets/0940-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        fruit_count = {}
        max_len = 0
        left = 0

        for right in range(len(fruits)):
            fruit_count[fruits[right]] = fruit_count.get(fruits[right], 0) + 1

            if len(fruit_count) > 2:
                fruit_count[fruits[left]] -= 1
                if fruit_count[fruits[left]] == 0:
                    del fruit_count[fruits[left]]
                left += 1

            if len(fruit_count) <= 2:
                max_len = max(max_len, right - left + 1)

        return max_len
        

File: 0940-fruit-into-baskets/0940-fruit-into-baskets.go
func totalFruit(fruits []int) int {
	mp := make(map[int]int)
	maxLen := 0
	left := 0

	for r := 0; r < len(fruits); r++ {
		mp[fruits[r]]++

		if len(mp) > 2 {
			mp[fruits[left]]--
			if mp[fruits[left]] == 0 {
				delete(mp, fruits[left])
			}
			left++
		}

		if len(mp) <= 2 {
			maxLen = max(maxLen, r-left+1)
		}
	}
	return maxLen
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}